### ISSUE TITLE
DOC: optimize.minimize: recipe for avoiding redundant work when calling `fun`, `jac`, and `hess` at same `x`

### ISSUE THREAD

**Issue by ahwillia** [scipy/scipy#9265]
Here is a simple optimization example (Poisson regression):

```python
# loss function and gradient
def f(b):
    Xb = np.dot(X, b)
    exp_Xb = np.exp(Xb)
    loss = exp_Xb.sum() - np.dot(y, Xb)
    grad = np.dot(X.T, exp_Xb - y)
    return loss, grad

# hessian
def hess(b):
    return np.dot(X.T, np.exp(np.dot(X, b))[:, None]*X)

# optimize
result = minimize(f, np.zeros(p), jac=True, hess=hess, method='newton-cg')
```

This works, but since `np.exp(np.dot(X, b))` is already computed inside `f(b)`, it would be more efficient and more readable to be able to do:

```python
# loss function and gradient
def f(b):
    Xb = np.dot(X, b)
    exp_Xb = np.exp(Xb)
    loss = exp_Xb.sum() - np.dot(y, Xb)
    grad = np.dot(X.T, exp_Xb - y)
    hess = np.dot(X.T, exp_Xb[:, None] * X) 
    return loss, grad, hess

# optimize
result = minimize(f, np.zeros(p), jac=True, hess=True, method='newton-cg')
```

But this doesn't seem supported, right now I get:

```
TypeError: 'bool' object is not callable
```

**Comment by ahwillia** [comment]
Here is a work-around that I will be using for now. This is a simple extension of [`MemoizeJac`](https://github.com/scipy/scipy/blob/v1.1.0/scipy/optimize/optimize.py#L53) wrapper defined internally in `scipy.optimize`.

```python
class ObjectiveWrapper:
    """
    Decorator that caches the gradient and hessian of the objective function.
    """
    def __init__(self, func):
        self.func = func
        self._grad = None
        self._hess = None
        self.x = None

    def __call__(self, x, *args):
        self.x = np.asarray(x).copy()
        fgh = self.func(x, *args)
        self._grad = fgh[1]
        self._hess = fgh[2]
        return fgh[0]

    def grad(self, x, *args):
        if self._grad is not None and np.alltrue(x == self.x):
            return self._grad
        else:
            self(x, *args)
            return self._grad

    def hess(self, x, *args):
        if self._hess is not None and np.alltrue(x == self.x):
            return self._hess
        else:
            self(x, *args)
            return self._hess
```

Then scipy.optimize.minimize is called like so:

```python
def f(x):
    # ... returns loss, gradient, hess as a tuple

obj = ObjectiveWrapper(f)
result = minimize(obj, x0, jac=obj.grad, hess=obj.hess, method='newton-cg')
```

It seems like this basic solution could be incorporated internally, perhaps even some simple modifications of `MemoizeJac` would work. I'm a bit hesitant to jump in and open a PR because I'm having a bit of trouble following the code at places. I'm sure I could figure it out, but maybe this would be really easy for someone already familiar with the internals? Let me know.

**Comment by mdhaber** [comment]
@antonior92 this is fixed in 1.2, right?

**Comment by ahwillia** [comment]
That's great! On a related note, is a similar option for `hessp=True` in the works?

**Comment by mdhaber** [comment]
@ahwillia I misread before. I actually don't think this is fixed for 1.2. @antonior92 thoughts?

**Comment by antonior92** [comment]
@mdhaber this is not implemented. You may have mistaken this for: #8883. Where I have fixed the incompatibility of ``trust-constr`` with the default interface.

@ahwillia I would like to point that computing the Hessian, together with the Jacobian and objective function is NOT necessarily more efficient than computing the Hessian in a separate function. Actually I think it will be more inefficiently most of the time. That is because, while you may end up reusing some information from previous computation, you also may end up computing the Hessian for situations where the Hessian would not be computed (most of the optimization methods evaluate the objective function in a lot of points where the Hessian is not needed). I think in the end it is cheaper to make an extra evaluation of the objective function than to evaluate the Hessian more times than needed.

**Comment by ahwillia** [comment]
Hmm yeah I see your point because you'll generally need some sort of line search with Newton's method. But in that case you'd also not need to recompute the gradient every time the objective is called. Maybe the `ObjectiveWrapper` approach that I mention above is really the best? In that case maybe we could explain this practice somewhere in the docs? Or do you think that there is only marginal performance improvements to be had along these lines?

**Comment by antonior92** [comment]
> Do you think that there is only marginal performance improvements to be had along these lines?

To be completely sincere, I think there is not much to be gained in term of performance along these lines. And I think it give the user the false impresion that he is actually getting a better performace while doing this

The way of really getting the best performance is something like:
```
class ObjectiveWrapper:
    def __init__(self):
        self.common_computations= None
        self.x = None

    def core_computations(self, x, *args):
        if not np.array_equal(x, self.x):
            # ...
            self.common_computations = common_computations
         return self.common_computation

    def func(self, x, *args):
        core_computations = self.core_computations(x, *args)
        # ...
        return func

    def grad(self, x, *args):
        core_computations = self.core_computations(x, *args)
        # ...
        return func

    def hess(self, x, *args):
        core_computations = self.core_computations(x, *args)
        # ...
        return func
```
But this would need to be done for your specific problem.

> But in that case you'd also not need to recompute the gradient every time the objective is called.

Yes, to be completely since I don't think the option ``jac=True`` is that great of a option neither

---------------

In summary, I think we would make the interface more complicated. It would give the users the impression that they would get some improvement in performance, while I don't think it will make much of a difference.

Maybe a best option would be to include in the documentation a paragraph or two explaining how to do the above trick I just mentioned

**Comment by ahwillia** [comment]
Thanks so much for helping me think this through. This has been really helpful.

> Maybe a best option would be to include in the documentation a paragraph or two explaining how to do the above trick I just mentioned.

I absolutely think this would be the best solution! Ideally we would also have a simple model that we can benchmark to precisely assess how much these reorganizations impact performance.

Should I close this issue and open a new one asking for better documentation on this issue? Or change the title of this one?

**Comment by antonior92** [comment]
> Thanks so much for helping me think this through. This has been really helpful.

Glad I've helped you out in this.

> Should I close this issue and open a new one asking for better documentation on this issue? Or change the title of this one?

I think it is better to just change the name of the issue accordingly.

**Comment by mdhaber** [comment]
@antonior92 quick question: `core_computations` in your example above should have `self.x = x` at the end, right?
